service/qall/roomstate_impl.go

```diff
@@ -0,0 +1,295 @@
+	r.RoomState = roomWithParticipants
+	return err
+}
+
+// AddParticipantToRoomState ルーム状態に参加者を追加
+func (r *Repository) AddParticipantToRoomState(room *livekit.Room, participant *livekit.ParticipantInfo) {
+	for i, roomState := range r.RoomState {
+		if roomState.RoomID.String() == room.Name {
+			t := time.Unix(participant.JoinedAt, 0).In(time.FixedZone("Asia/Tokyo", 9*60*60))
+			r.RoomState[i].Participants = append(r.RoomState[i].Participants, Participant{
+				Identity:   &participant.Identity,
+				JoinedAt:   &t,
+				Name:       &participant.Name,
+				Attributes: &participant.Attributes,
+				CanPublish: &participant.Permission.CanPublish,
+			})
+
+			if r.Hub != nil {
+				r.Hub.Publish(hub.Message{
+					Name: event.QallRoomStateChanged,
+					Fields: hub.Fields{
+						"room_id": roomState.RoomID,
+						"state":   &r.RoomState[i],
+					},
+				})
+			}
+
+			break
+		}
+	}
+}
+
+// UpdateParticipantCanPublish 参加者の発言権限を更新
+func (r *Repository) UpdateParticipantCanPublish(roomdID string, participantID string, canPublish bool) {
+	for i, roomState := range r.RoomState {
+		if roomState.RoomID.String() == roomdID {
+			for j, participant := range roomState.Participants {
+				if *participant.Identity == participantID {
+					r.RoomState[i].Participants[j].CanPublish = &canPublish
+
+					if r.Hub != nil {
+						r.Hub.Publish(hub.Message{
+							Name: event.QallRoomStateChanged,
+							Fields: hub.Fields{
+								"room_id": roomState.RoomID,
+								"state":   &r.RoomState[i],
+							},
+						})
+					}
+
+					break
+				}
+			}
+			break
+		}
+	}
+}
+
+// UpdateParticipant 参加者情報を更新
+func (r *Repository) UpdateParticipant(roomdID string, participant *livekit.ParticipantInfo) {
+	for i, roomState := range r.RoomState {
+		if roomState.RoomID.String() == roomdID {
+			for j, p := range roomState.Participants {
+				if *p.Identity == participant.Identity {
+					t := time.Unix(participant.JoinedAt, 0).In(time.FixedZone("Asia/Tokyo", 9*60*60))
+					r.RoomState[i].Participants[j] = Participant{
+						Identity:   &participant.Identity,
+						JoinedAt:   &t,
+						Name:       &participant.Name,
+						Attributes: &participant.Attributes,
+						CanPublish: &participant.Permission.CanPublish,
+					}
+
+					if r.Hub != nil {
+						r.Hub.Publish(hub.Message{
+							Name: event.QallRoomStateChanged,
+							Fields: hub.Fields{
+								"room_id": roomState.RoomID,
+								"state":   &r.RoomState[i],
+							},
+						})
+					}
+
+					break
+				}
+			}
+			break
+		}
+	}
+}
+
+// RemoveParticipant ルームから参加者を削除
+func (r *Repository) RemoveParticipant(roomdID string, participantID string) {
+	for i, roomState := range r.RoomState {
+		if roomState.RoomID.String() == roomdID {
+			for j, participant := range roomState.Participants {
+				if *participant.Identity == participantID {
+					r.RoomState[i].Participants = append(r.RoomState[i].Participants[:j], r.RoomState[i].Participants[j+1:]...)
```

Comment:
ここは以下のような書き方で簡略化できそうです
```suggestion
					r.RoomState[i].Participants = slices.Delete(r.RoomState[i].Participants, j, j+1)
```

Reply:
ここなんですが、 `slices` package が必要なのでこれを適用したままだとビルドが落ちちゃう気がします 😓

Reply:
ToDo: コード触るときにpackage追加する